lib/ansible/modules/remote_management/redfish/redfish_facts.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2017-2018 Dell EMC Inc.
+# GNU General Public License v3.0+ (see LICENSE or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'metadata_version': '1.1'}
+
+DOCUMENTATION = '''
+---
+module: redfish_facts
+version_added: "2.7"
+short_description: Manages Out-Of-Band controllers using Redfish APIs
+description:
+  - Builds Redfish URIs locally and sends them to remote OOB controllers to
+    get information back.
+  - Information retrieved is placed in a location specified by the user.
+options:
+  category:
+    required: true
+    description:
+      - Action category to execute on server
+    choices: ["Inventory", "Accounts", "System", "Update", "Manager", "Chassis"]
+  command:
+    required: true
+    description:
+      - Command to execute on server
+  baseuri:
+    required: true
+    description:
+      - Base URI of OOB controller
+  user:
+    required: true
+    description:
+      - User for authentication with OOB controller
+  password:
+    required: true
+    description:
+      - Password for authentication with OOB controller
+
+author: "Jose Delarosa (github: jose-delarosa)"
+'''
+
+EXAMPLES = '''
+  - name: Getting system inventory
+    redfish_facts:
+      category: Inventory
+      command: GetSystemInventory
+      baseuri: "{{ baseuri }}"
+      user: "{{ user }}"
+      password: "{{ password }}"
+
+  - name: Get CPU Inventory
+    redfish_facts:
+      category: Inventory
+      command: GetCpuInventory
+      baseuri: "{{ baseuri }}"
+      user: "{{ user}}"
+      password: "{{ password }}"
+
+  - name: Get BIOS attributes
+    redfish_facts:
+      category: System
+      command: GetBiosAttributes
+      baseuri: "{{ baseuri }}"
+      user: "{{ user }}"
+      password: "{{ password }}"
+
+  - name: List all users
+    redfish_facts:
+      category: Accounts
+      command: ListUsers
+      baseuri: "{{ baseuri }}"
+      user: "{{ user }}"
+      password: "{{ password }}"
+'''
+
+RETURN = '''
+result:
+    description: different results depending on task
+    returned: always
+    type: dict
+    sample: List of CPUs on system
+'''
+
+import os
+import json
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.redfish_utils import RedfishUtils
+
+
+def main():
+    result = {}
+    module = AnsibleModule(
+        argument_spec=dict(
+            category=dict(required=True, choices=["Inventory", "Accounts", "System", "Update", "Manager", "Chassis"]),
```

Comment:
Set a default value for category to allow this module to work without any parameters being passed in. This provides a better UI and will fit with future plans we have for allowing a user to specify a list of facts modules to be run by default.

Reply:
Making this field a list rather than a string would allow the user to specify which categories they would like gathered. Having Ansible run one task and gather multiple sets of facts is more efficient from an Ansible perspective than calling the fact module multiple times with a different category each time. Even if you have to construct multiple queries to the API, this is still more efficient than multiple tasks being run.

Reply:
Each invocation requires a category/command pair. Setting a default value for category would not be efficient because each category deals with very different sets of data: HW inventory, FW inventory, users, BIOS attributes, etc. Which one is considered the default? Setting a default value for command is more straight forward, as each category has a basic sets of facts that can be considered "default".

Reply:
Ok, I see a bit more clearly now what's going on. That does make things trickier.

Would it be possible to remove some of this granularity for a simpler UI? For instance, only allow specifying the categories and the list of commands to be run are pre-configured in the module and cannot be changed. Run the most common (and least expensive time/performance wise) commands by default for each category, but have an option for running _all_ the commands for each category.

Maybe create a mapping of categories to commands to run, and pick the mapping based on a `gather_all` parameter.

```python
CATEGORY_COMMANDS_ALL = {
    "Inventory":  ["GetSystemInventory", "GetPsuInventory", "GetCpuInventory", "GetNicInventory"],
    "Accounts":  ["ListUsers"],
    "System":  ["GetBiosAttributes", "GetBiosBootOrder"],
    "Update":  ["GetFirmwareInventory"],
    "Manager":  ["GetManagerAttributes", "GetLogs"],
    "Chassis": [],
}

CATEGORY_COMMANDS_DEFAULT = {
    "Inventory":  ["GetSystemInventory"],
    "Accounts":  ["ListUsers"],
    "System":  ["GetBiosAttributes"],
    "Update":  ["GetFirmwareInventory"],
    "Manager":  ["GetManagerAttributes"],
    "Chassis": [],
}
```

If this won't work, we can look at some other options. I'm also not sure if this is too limiting because the API would add new commands that wouldn't exist in the module.